### PR TITLE
Blocklist Sync

### PR DESCRIPTION
**Pull request by peterjan**
# PULL REQUEST

## Overview
This PR adds a new feature that synchronizes the blocklist of predefined portals with the local skyd instance. This allows new portals to bootstrap from other portals' blocklists very easily. It's important to note that the hashes added from other portals are only added to the local skyd block list and are not added to the database as a blocked Skylink. I believe that to be the desired behavior seeing as we don't have a reporter and there's really no need for all of those databases to sync up this way, we only care that the hash is on the blocklist.

## Example for Visual Changes
N/A

## Checklist
Review and complete the checklist to ensure that the PR is complete before assigned to an approver.
 - [x] All new methods or updated methods have clear docstrings
 - [x] Testing added or updated for new methods
 - [ ] Verify if any changes impact the WebPortal Health Checks
 - [ ] Approriate documentation updated
 - [ ] Changelog file created

## Issues Closed
N/A

### CODE REVIEW

File: skyd/api.go
Comment (ChrisSchinnerl) [inline skyd/api.go:76]: This seems like it should use our client library because it's just a lightweight reimplementation of it which doesn't check the status code and uses a bunch of hardcoded values.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:27]: This seems like a bit too much considering that we have had providers request that we block certain content within an hour. I do understand that we don't like fetching that much data at once though.

So one thought might be to extend skyd itself to also provide a timestamp of the last time we updated the blocklist. Then you poll that more frequently and only if it changed we fetch the whole thing again.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:55]: "no" might be better than "invalid" in this case.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:109]: This is pretty futile. Because we are not using a waitgroup in this function. So nothing is waiting for the thread to finish which in turn means it will be killed as soon as main exits.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:93]: I'd pull this out into a `threadedSyncLoop` to reduce the x-winginess.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:98]: This loop could also be a separate `syncPortalsWithSkyd` method.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:124]: Same as before. A lot of duplicate http code which does the exact same stuff. Create request, send request, defer close body, and decode. Might as well just use the skyd client to reduce that to a few lines.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:87]: Why are we only ever syncing the db once? Is there no chance it gets updated in the meantime? Or do we have another module in the blocker which updates skyd from the db? In that case the architecture would be a bit chaotic imo.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:82]: There doesn't seem to be a test for the actual syncthread picking up on changes.

File: syncer/syncer.go
Comment (ChrisSchinnerl) [inline syncer/syncer.go:208]: I don't think there is testing for this method.

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]

File: database/database.go
Comment (peterjan) [reply database/database.go:247]: it's used by the `syncer` module when we sync the database with skyd on startup

File: main.go
Comment (peterjan) [reply main.go:181]: how so? you mean because we're re-adding the prefix later? I don't mind that personally, it's straightforward:

It's 
```go
       portalURL = strings.TrimSpace(portalURL)
	portalURL = strings.TrimPrefix(portalURL, "https://")
	portalURL = strings.TrimPrefix(portalURL, "http://")
	portalURL = strings.TrimSuffix(portalURL, "/")
	return fmt.Sprintf("https://%s", portalURL)
```
vs
```go
       portalURL = strings.TrimSpace(portalURL)
	portalURL = strings.TrimSuffix(portalURL, "/")
	if strings.HasPrefix(portalURL, "https://") {
		return portalURL
	}
	portalURL = strings.TrimPrefix(portalURL, "http://")
	return fmt.Sprintf("https://%s", portalURL)
```

I'd take the first version any day, but I'll update I don't really care.

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:27]: I think I should've done a better job explaining the idea in the PR. The idea for the blocklist sync is not to keep up to date with the local skyd instance, but rather to bootstrap a portal with the blocklist of another portal. So really nothing changes and the portal operator should still run the abuse software to ensure blocking happens immediately, which it will.

This 4 hour sync time is merely an interval with which we fetch the blocklist from another portal. The idea of adding a timestamp to the blocklist endpoint though is great, we should definitely do that. I just don't see the issue with the 4h interval here. Maybe you don't see one either now that it's more clear what the feature is/does?

File: blocker/blocker.go
Comment (peterjan) [reply blocker/blocker.go:114]: updated, thx

File: database/database.go
Comment (peterjan) [reply database/database.go]: There used to be but in the new architecture, there's no reason for them to be sorted. For what it's worth though, if you're sorting on an indexed field (which is the case here) in mongo the overhead is really very very minimal because it can just use its index which is sorted already.

That said, I was able to remove it because we got rid of the `latestBlockTime` timestamp thing.

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go]: restructured

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go]: restructured to be less chaotic, this was actually made redundant by the changes I was planning to make already, when writing the syncer I just didn't realize this was not necessary any longer

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:82]: added

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go]: added

File: main_test.go
Comment (peterjan) [reply main_test.go:12]: added

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:116]: resolving in favor of having one comment on this issue

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:111]: is this F/U material? definitely a good suggestion though

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:27]: I thought about this more and I have come to see your point. I think this is a F/U issue as it needs an update to skyd but it's definitely one that we want to make. After that we'd want to indeed lower this to something like every 15 minutes or so I'd say.

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:101]: True, was just copying the pattern that was there in the blocker. I'll update this for it to make sense and introduce a waitgroup. Do you mind if this is done in a F/U? I'll definitely F/U with this soon, just want to avoid bloating the PR further.

File: syncer/syncer.go
Comment (ro-tex) [inline syncer/syncer.go:152]: This method downloads blocklists from other portals and then blocks them on the local server but it doesn't add them to the portal's blocklist itself. 

If we only block the hashes in skyd but don't put them in the DB we have a few suboptimal outcomes:
1. Each new server that you add to your cluster must rely on syncing those blocklist entries from the outside server as well, instead of fetching them from the local DB.
2. You can't get a complete list of what's blocked on your portal. Knowing what's blocked might be useful if you need to argue with law enforcement or unblock something due to a valid challenge of the original block.
3. If you're running your own malware-scanning service, you'll end up re-downloading and re-scanning something somebody already scanned.
4. If you whitelist some legitimate content that was erroneously blocked, those blocks will be dropped from your local `skyd`, only to be reinstated on the next pull from an external server which still has the illegitimate block.

All in all, I don't see a benefit in not updating our own ledger of bad skylink hashes and blindly accepting what other portals have blocked without double-checking with our own list first.

File: blocker/blocker.go
Comment (peterjan) [reply blocker/blocker.go:50]: That's not that much of an issue. On server restart we fetch the skyd blocklist and all hashes from the database, we diff them and we add what's missing. So this was intentional and not an issue IMO. I want to avoid persisting this value.

File: main_test.go
Comment (peterjan) [reply main_test.go:182]: good point, updated!

File: blocker/blocker.go
Comment (peterjan) [reply blocker/blocker.go:122]: what do you propose? I personally don't mind the naming here

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:152]: I see your point but this essentially puts into question the entire PR. I personally don't think any of the bullet points above are an issue really, but others might feel differently. The reason I did not add the hashes to the local database is that those records would be totally different from the other ones. There is no reporter, there are no tags, there is no Skylink and so on, it's just a hash. The sync is just a way for new portal operators to bootstrap their blocklist with the blocklist of a trusted portal and do so with ease.

1. Every server should indeed have the same portals defined in the environment for that local server's skyd to contain all the hashes another server would have. I don't think this is an issue really, we manage env variables from a single source of truth anyway.
2. You can get a list of blocked hashes from skyd. We are removing away from storing skylinks everywhere anyway so pulling a list of skylinks you are blocking is impossible.
3. I don't fully understand this bullet point, but you probably wouldn't want to trust the scan results of a 3rd party anyway.
4. Currently the allow list is based on skylinks. That's being moved in the second part of the skylink to hash refactor. When that is complete I will be able to check the hashes against the whitelist and check before blocking, but this bullet is a good point.

This feature is just the result of Matt saying "it would be nice to have a button somewhere to sync the blocklist from portal A to portal B". I think there are plenty of use cases where as a portal operator you would be fine with syncing the blocklist from a trusted friend portal, and if you're not you simply don't do it.

File: blocker/blocker.go
Comment (ro-tex) [reply blocker/blocker.go:122]: Even `MarkAsNotFailed` would be better, as it won't imply the existence of a `succeeded` field or something of the sort.

File: syncer/syncer.go
Comment (ro-tex) [reply syncer/syncer.go:152]: 2. You can get a list of blocked hashed *per server*, sure, but you'd need to query each server separately and then merge the results. Anywhere where I'm talking about skylinks I mean their hash representation.
3. The idea here is that we are already blocking some skylink because it's been blocked by a portal we sync with. But since it's been requested and it's not in the DB, we will still run it through our malware-scanner, wasting resources. And the best part here is that regardless of the result of the scan, the hash will still be blocked anyway, because we don't check those hashed against our allowlist.

I've heard this feature described as "a way to bootstrap a portal" and I think that's fine. But if it's means as a bootstrapping tool then:
1. It shouldn't be run routinely, keeping in sync with the other portal.
2. It should still store the hashes in the DB because that's what bootstrapping usually means - getting a solid starting point on which to build. If we don't do that and we later add another server to our portal but the portal we synced from is down or dead, the new server will have a different blocklist from the rest of our portal. That's not ideal from the standpoint of predictable behaviour of our own servers.

File: blocker/blocker.go
Comment (ChrisSchinnerl) [reply blocker/blocker.go:122]: I think it's fine. Because it actually succeeded. Sure the field says `failed` but the point of the verbose setter is to move the argument into the method name as opposed to `MarkFailed(true || false)`. Which I would be fine with as well.

Although if I had to nitpick I'd probably prefer a shorter name like `MarkSucceeded` and `MarkFailed`.

Only nit I'd probably have is that it could be shortened to `MarkFailed` and `MarkSucceeded`.

Comment (peterjan) [reply blocker/blocker.go:122]: Ok, I'll update it to `MarkSucceeded` and `MarkFailed`. I would usually entertain the reviewer and update it but in this case, I just disagree that it's confusing. There's no `succeeded field because a boolean should not be defined by two fields in the database. `MarkSucceeded` is the right name because that's just what it does, _how_ it marks the document as succeeded is irrelevant. I owe you one Ivo :)

File: syncer/syncer.go
Comment (peterjan) [reply syncer/syncer.go:152]: I can't update the PR in a way that would allow me to resolve your comment. Adding them to the database from the blocker is not the desired approach in my opinion because we would not have a reporter, nor a list of tags. If we would follow your reasoning we would need an entirely different tool that imports the entire mongo collection or keeps those in sync. I'll wait for comments from either @MSevey or @ChrisSchinnerl (or both) to proceed.

Comment (ChrisSchinnerl) [reply syncer/syncer.go:152]: I think we need to establish what we want to achieve first.
1. A one-time bootstrap - in that case this PR may be suboptimal because we don't bootstrap our own db.
2. A way to keep up-to-date with a trusted portal's list. Which will usually be used by people who don't feed their own db and just want to mirror a different portal.

Comment (peterjan) [reply syncer/syncer.go:152]: Matt asked for a one-time bootstrap button. I just figured I was extending the idea in a good direction. 

I felt syncing > bootstrapping, because the end result will be that if you configure a sync, your portal will benefit from having abusive skylinks blocked, even though they were not reported to you directly. In the end, we want to just block CSAM links on as many portals as fast as possible right.

Comment (peterjan) [reply syncer/syncer.go:152]: ok resolved on Discord - we'll add a route to the blocker (partially) exposing the blocklist collection
instead of syncing with a server's blocklist, we'll sync with a portal's blocklist, which is way better
and I fully agree with merging those db entities

Comment (peterjan) [reply syncer/syncer.go:152]: marked as draft as I need to make some changes, resolving this comment

File: database/database.go
Comment (peterjan) [inline database/database.go:195]: the naming is off here but I'm tackling that in another PR that runs through the whole blocker and gets rid of the term "skylink" where its not appropriate any longer

File: api/client.go
Comment (ChrisSchinnerl) [inline api/client.go:58]:
```go
defer func() {
  // drain and close body just in case
  _ = io.Copy(ioutil.Discard, resp.Body)
  _ = resp.Body.Close()
}

File: database/database.go
Comment (ChrisSchinnerl) [inline database/database.go:217]: could you explain that this makes it so a single write can fail but that doesn't prevent the other ones from being written?

File: database/database.go
Comment (ChrisSchinnerl) [inline database/database.go:219]: Just to verify. When I write 3 entries. One of them is a duplicate. Another one failed because of a different reason. 
Then this will not print the error for the entry that failed without the duplicate error right? We also won't return an error here. So as long as there is a single duplicate key error here it overshadows any other error that might happen.

File: database/skylink.go
Comment (ChrisSchinnerl) [inline database/skylink.go:102]: `bsl.TimestampAdded.IsZero()`

File: skyd/api.go
Comment (ChrisSchinnerl) [inline skyd/api.go:112]: do you need this? Aren't the hashes marshalled to strings by `json.Marshal` already? Their `json.Marshaler` implementation might even use the `.String()` method.

File: skyd/api.go
Comment (ChrisSchinnerl) [inline skyd/api.go:130]: Might warrant a generic `.post` method like we are already doing for get requests.

File: skyd/api.go
Comment (ChrisSchinnerl) [inline skyd/api.go:143]: Same as before. Defer also flushing the body and then use a `Decoder` instead

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]

File: database/skylink.go
Comment (peterjan) [reply database/skylink.go:102]: done

File: api/client.go
Comment (peterjan) [reply api/client.go:67]: updated, thanks

File: api/client.go
Comment (peterjan) [reply api/client.go:58]: done

File: database/database.go
Comment (peterjan) [reply database/database.go:217]: done

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:143]: done

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:130]: I want to get rid of this entire file and move it to the client. I already have a bunch of that in my stash but I stashed it because of the size of the PR. That's probably the last PR I want to make on the blocker (after the ones that are open are merged). Seeing as I was planning on doing this already I'll do this in a F/U.

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:112]: I think I do need that because `SkynetBlocklistPOST` expects a `[]string`...

```
	SkynetBlocklistPOST struct {
		Add    []string `json:"add"`
		Remove []string `json:"remove"`

		// IsHash indicates if the supplied Add and Remove strings are already
		// hashes of Skylinks
		IsHash bool `json:"ishash"`
	}
```

File: database/database.go
Comment (peterjan) [reply database/database.go:219]: Ok, point taken. Not an issue! I can simply cast it to the interface and range over all write errors

File: database/database.go
Comment (peterjan) [reply database/database.go:359]: it's just a little warning to devs to be vigilant when change the `filter` below and change `"$ne": true` to `$eq: false`... that's important for when the field is undefined

File: syncer/syncer_test.go
Comment (peterjan) [reply syncer/syncer_test.go:90]: interesting! I'll check that out for sure, will keep this as is for the time being though to get passed this  review

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:143]: this is because `skyd/api.go` will be replaced by the client, I'm planning to resolve this [comment](https://github.com/SkynetLabs/blocker/pull/19#discussion_r819443292) in an issue for that

File: database/database.go
Comment (ro-tex) [reply database/database.go:359]: So, `$eq: false` requires the field to be set and won't match it if it's not set, while `$ne: true` will match an unset field. Got it.

Comment (peterjan) [reply database/database.go:359]: right

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go:130]: fine with that?

File: database/database.go
Comment (ro-tex) [inline database/database.go:467]: While this does ignore all `DuplicateKey` errors, it also truncates the list of errors to the first one it encounters.
That is almost OK because any non-dup error will result in the current portal sync being aborted. But it might not give us all errors that occurred. Is that important? If not, feel free to resolve this.

**Review by ro-tex — APPROVED** [review]

File: database/database.go
Comment (peterjan) [reply database/database.go:467]: It actually simply returns the original error as soon as it finds a write error that's not a duplicate key error. But even if it weren't it's probably okay. This is an edge case that I don't expect to occur ever, and even if it does it's still not much of an issue.

**Review by ChrisSchinnerl — APPROVED** [review]

**Review by ro-tex — APPROVED** [review]

File: skyd/api.go
Comment (peterjan) [reply skyd/api.go]: done

**Pull request merged**
